### PR TITLE
Submodule status: Throttle instead of drop frequent updates

### PR DESCRIPTION
**Pull request by gerhardol**
I believe this is ready, but did not plan to include it in 3.4
Submitting now for discussion of #8049 

## Proposed changes

SubmoduleStatusProvider will not update the status more often than every 15th second to limit load.
This was added before the status update was fully async and before updates were pruned (latest change in #8116).
The status should not be noticable (at least comparable to git-status, that can take a long time in big repos with many submodules).

The update limitation remains with this PR, but instead of dropping frequent updates, the handling is throttled to update no more often than every 15th second.
The handling has also been simplified.

A change to specific mention is that previously, if not the top module was selected, the complete structure including the current submodule was updated. The current submodule was then updated a second time. (There was a comment that the update not triggered by git-status would skip this, but that was incorrect.)
This will maybe be noticeable on a deep tree where a module below submodule is updated. (but very minor effect)

## Test methodology 

Tests exists, a test is disabled for the same reason as before.
Can maybe be improved.

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by gerhardol** [comment]
Squashed and rebased https://github.com/RussKie/gitextensions/tree/fix_8077 on master, rebased this on that commit and fixed the commented out tests, added a few lines more
This was a problem with the datastructures

The bug in the application is not too severe. The status for submodules if not the top project was checked out was incorrect. So the problem requires at least three levels of submodules.

The drop down looked like this:
image

With the correction, it looks like the following (sidepanel is correct)
image

**Comment by gerhardol** [comment]
The current implementation can be further simplified, no need to separate topmodule and submodule structure update

File: GitCommands/Submodules/SubmoduleStatusProvider.cs [hunk 1]
```diff
@@ -136,24 +117,29 @@ public async Task UpdateSubmodulesStructureAsync(string workingDirectory, string
...
+                await Task.Delay(TimeSpan.FromMilliseconds(300), cancelToken);
+                cancelToken.ThrowIfCancellationRequested();
             }
 
-            var cancelToken = _submodulesStatusSequence.Next();
-            _gitStatusWhileUpdatingStructure = null;
-
-            await TaskScheduler.Default;
+            // Throttle updates triggered from status updates
+            TimeSpan remaining = _previousSubmoduleUpdateTime - DateTime.Now.AddSeconds(-15);
```
Comment (RussKie) [inline GitCommands/Submodules/SubmoduleStatusProvider.cs:137]: Suggest defining a const and use it here. Makes it easier to track.

**Comment by gerhardol** [comment]
I submitted this to support #8049 in a safe way, but it has maybe grown too big for 3.4
I would like to add some more test cases for changing revisions as well (could be done later too).
I may submit an alternative with a few line changes for 3.4 instead.

**Comment by gerhardol** [comment]
Testcases added, also checking AllSubmodules/OurSubmodules (what failed in the old testcases)
I added 300 lines of testcases, now many more but not all usages are covered. So it is still not a safenet. (Also before reviews, I am blind for issues myself now.)
Added some tests for the actual working, even if it is incomplete. For instance, if not a top module is checked out, then git-status run on the submodule and can determine that the top module is dirty. However, clearing the dirty file cannot determine that the top module is no longer dirty.

**Comment by mstv** [comment]
> ChangedFiles does not report the expected number of changes in Appveyor

Without having looked at the code, it might help to add
`await AsyncTestHelper.JoinPendingOperationsAsync(AsyncTestHelper.UnexpectedTimeout);`
and `forceUpdate: true`.

File: GitCommands/Submodules/SubmoduleStatusProvider.cs [hunk 2]
```diff
@@ -136,24 +117,30 @@ public async Task UpdateSubmodulesStructureAsync(string workingDirectory, string
...
+                await Task.Delay(TimeSpan.FromMilliseconds(300), cancelToken);
+                cancelToken.ThrowIfCancellationRequested();
             }
 
-            var cancelToken = _submodulesStatusSequence.Next();
-            _gitStatusWhileUpdatingStructure = null;
-
-            await TaskScheduler.Default;
+            // Throttle updates triggered from status updates
+            const int minRefreshInterval = 15;
```
Comment (mstv) [inline GitCommands/Submodules/SubmoduleStatusProvider.cs:137]: I guess, Igor meant to create a class constant, which is visible at the top of the file.

**Comment by gerhardol** [comment]
Split tests and added a few more. Six tests fails on AppVeyor though. It is consistent behavior though, no evident race condition.
I check what Git options Appveyor used, nothing strange (and GE should override user settings anyway if needed.)
Will squash this (it is test cases that has changed since the review).

@mstv any more hints to why the tests fail?
Anything specific to be done so the testcases are not cleaned up and access with RDP?

**Comment by mstv** [comment]
> Six tests fails on AppVeyor though.

I have not found a regarding failed AppVeyor run.

> Will squash.

OK for me.

**Comment by mstv** [comment]
Printf debugging still is not the worst idea.
You could attach the debugger to the test runner.
(IIRC, the AppVeyor VM has only one cpu core with hyperthreading.)

**Comment by gerhardol** [comment]
Eventually fixed the problem, same problem as #7861: AppVeyor does not set an email address which causes commit to fail.
The testcases sets email addresses in the local config but when adding a module as a submodule, the local config is not copied.
I intend to squash the two first commits.

**Comment by gerhardol** [comment]
One more corner case, with another test case

**Comment by gerhardol** [comment]
@msftbot merge in 24 hours
